### ISSUE TITLE
Remove Room Inheritance

### ISSUE THREAD

**Issue by iampremo** [YoYoGames/GameMaker-Feature-Requests#1359]
### Is your feature request related to a problem?

Room Inheritance has many fundamental issues with its functionality, complicates general development of features and the Room Editor, and there are far better ways of achieving the same results.

### Describe the solution you'd like

Remove the functionality of room inheritance.

**Comment by rwkay** [comment]
We are adding this Feature Request ourselves as we would like to remove Room Inheritance going forward... this is your opportunity to tell us what you are actually using it for, so we can work out what needs to be done to replace it.

**Comment by Alphish** [comment]
Personally, I never used it, so it wouldn't really affect my workflow that much. Might also have something to do with the fact that Room Inheritance had a reputation for being broken, so I didn't really touch it. An alternative would be polishing up the Room Inheritance so that it's actually usable, so either fix it or ditch it.

If it's removed, then the project converter tool will need to propagate inherited data to rooms themselves, so that no existing data is lost.

Also, I guess it would be good to figure out what makes people want to use room inheritance in the first place, and whether these needs are properly addressed by an alternative.

One thing I can think of is having a consistent set of layers, which then can then be used to place elements via Room Editor; when it comes to that, an alternative such as e.g. building layers from code won't address placing things via Room Editor. Another option would be adding layers manually - on paper the time spent on adding layers should be negligible compared to the time spent placing instances and the like. But manual adding is more prone to typos, which can be especially troublesome if layer names are used in code for further processing.

**Comment by HannulaTero** [comment]
Although the idea for inherited rooms is fine, I think overall it is confusing, as rooms have so many moving parts. 
There are some use-cases for inherited rooms, but one could live without them with building just single room, where all base layout is done, and then having separate layers which are deactivated/ activated according to the need. Like start of the room deactivate all instance layers which are "act specific", and then only activate whatever is required.

**Comment by phablix** [comment]
I would like a rework or remove and replace it in a future with a better system.
Something that would let us control in the parent room how each parent layer should behave in the child rooms would be great.

For example to control if a layer should inherit the instances that are there or not, and only inherit the layer name or the depth. 

This could work for organization purposes, and when we need for example to rename a layer in the parent, all child layers would be renamed too. And we wouldn't need to rename or change the depth of layers on tons of child rooms manually...

**Comment by JujuAdams** [comment]
☠️ Never seen it used effectively

**Comment by Totobal5** [comment]
Remove it...

**Comment by TheNocturne** [comment]
Ummmm... I'm kinda torn on this one. I know it's quite buggy right now, but my current long-term project absolutely RELIES on it! So removing/changing it in the near future (ie: the next 6-12 months) would seriously impact my project. Yes, I can stay on the last version that still has it, but with the way the team is adding new features and functions to GameMaker, I'd rather not do that and miss out on all the good stuff!

This, however, is a very selfish attitude and - in general - I think the feature was never correctly implemented and has several bugs/edge cases that make it difficult use predictably. So, if the best way forward is to remove it and replace it with something new, then I guess that's fine. It's potentially a very powerful concept (and the present limited iteration has certainly made my life easier on my current project), so I don't think it should _just_ be removed. Something "better" should take it's place...

That said, if you can simply fix the current implementation then so much the better, imho, but I have a vested interest in it not changing too much... 😅

**Comment by rwkay** [comment]
Current thoughts are too leave it in the current room editor but in the New Room Editor (which we are designing / thinking about now) we are thinking about removing it completely.

You would be able to use the current room editor for your current project.

**Comment by TheNocturne** [comment]
> Current thoughts are too leave it (Room Inheritance) in the current room editor but in the New Room Editor (which we are designing / thinking about now) we are thinking about removing it completely.
> 
> You would be able to use the current room editor for your current project.

Interesting. This sounds like a good way to go and definitely gets my support.

**Comment by mycelia** [comment]
I used it extensively in my last project but I think it just made things more complicated. Being able to copy layers, camera settings etc easily from one room to another would be more useful.

**Comment by Mercerenies** [comment]
I gave room inheritance a shot last year. It's never quite worked right, and for every project since, I've simply had a "prototype" room that gets copy-pasted every time I need a new level. In the ideal world, I could see the feature doing some good if (a) all the bugs were squashed, and (b) the UX was changed to make it far more intuitive, especially to beginners. But given the amount of effort that would take, I'm in support of removing the feature going forward and calling it a well-intentioned but failed experiment.

**Comment by JeffersonJales** [comment]
I kinda like the ideia of Room Inheritance, but the current state of the feature is too buggy.

> Also, I guess it would be good to figure out what makes people want to use room inheritance in the first place, and whether these needs are properly addressed by an alternative.

I've been using it for creating cutscenes. 
It's made a lot easier for the rest of the group to just inherent the template room and setup the cutscene behavior.
But I think with the upcomming Prefabs and GUI Layers, the Room Inheritance will be obsolete for real.

**Comment by FoxyOfJungle** [comment]
I personally never need to use Room Inheritance, so I'm fine without it. There are better ways to deal with the situation, which for me make me feel like I have more control over things, which is what I particularly prefer.

**Comment by dev-dwarf** [comment]
Kill it! I think a much more useful feature would be putting rooms inside of other rooms. This would be similar to a composition vs inheritance tradeoff. That would also be a big ask tho..

**Comment by XorDev** [comment]
As others are saying, the current implementation of room inheritance is unreliable.
That being said, I believe the concept is close to gold. There are many occasions where I'd like have a common view size across all my rooms, perhaps a background layer or in the future, a shared GUI layer. This means that any change to the parent room, will be inherited to all the child rooms, allowing for resolution and other changes to be done quickly. This is what I'd like to see from room inheritance. A simple way to change parameters across multiple rooms at once.
I could nested parent rooms being useful also. For example, a level game where you might have a parent room for all the levels to set a common resolution/view. Then you could have a child room, which is parent to levels 0-9, giving them a particular theme and background, and another parent for levels 10-19 and so on. Could make development easier for these such games.

**Comment by KormexGit** [comment]
I stopped using room inheritance due to how buggy it is, and now I use a template room that I duplicate instead. However, it'd be nice to have some way to edit layers for all rooms at once, since right now if I ever want to add/remove/rename any layers, I have to go back and do it in all my existing rooms one by one. I imagine this sort of mass layer editing could be added in a way that's not inheritance, so I'd like to see that. Something like how you can select multiple assets and edit properties for all of them at once in the inspector, but for rooms

**Comment by mmMarzban** [comment]
I have a very bad experience using it. It created a ton of issues and problems for me and therefore I decided not to use it at all for any other project.

**Comment by DragoniteSpam** [comment]
I feel like all of the things room inheritance was originally intended to allow you to do (creating user interface layers shared between rooms, template rooms, etc) now have better solutions elsewhere.

**Comment by AtlaStar** [comment]
If rooms had event callbacks and had their own struct which we could access via dot access on the room, I would argue that room inheritance is needed...but I am also of the mindset that things like layers should be their own type that is untethered from the room, and that cameras and views should also be stored in a way that is not attached to rooms before even entertaining the idea of rooms being more object like.

So I agree, the feature should be gotten rid of and maybe revisited in the future if rooms become more like objects/structs

**Comment by DAndrewBox** [comment]
I didn't used it because it is too buggy to begin with

Maybe just change it to "template rooms" that don't inherit anything more than the "structure" of the room, so i.e:

Have a template room with all the settings, objects and layers I want to use on children rooms so i can just press a button like "Create room from template" and have it all set.

Nothing about inherance of IDs or parent rooms with somewhat same ID, just a new room with new asset IDs, new layer IDs, etc.

I know this is something similar to what we already have, but maybe removing some features that make this system so buggy will be a better solution than removing it

**Comment by muddrox** [comment]
While I agree that room inheritance is borderline useless in its current decrepit state, I think it would be extremely useful if it were stable and complete.  It just feels unfinished in its current state.  If there are no plans to fix and improve it then sure, remove it, but I think it's a shame to lose it without anything to effectively replace it.  My vote is to fix it and improve its utility.

**Comment by NathanWolf** [comment]
I am currently using room inheritance, but I'm pretty new to Gamemaker. I am mostly using inheritance for shared UI layers, but from what others have posted here, it sounds like there is probably a better way (maybe just persistent objects created in the first room? I'm not sure).

I do really like the idea of room inheritance, but unless there were some fixes (or at least an easy way to force re-inherit on all rooms or children of a specific room) I will admit it's not very useable. I would say that's the only real feature that is lacking for me, some way to update children after making changes to a parent, which is currently a very tedious process when you have a lot of rooms.

I think removing the feature for the new room editor is a great compromise, it'd be nice to avoid breaking existing projects. Once I learn a little more I'm sure I could transition my projects away from room inheritance, and others could likely do the same.

**Comment by Gamer-XP** [comment]
I've tried using it at some point to hold layers with backgrounds and default room settings. But there were a lot of bugs because fo it so I hade to drop it.
Instead of room inheritance I'd prefer if there were something like prefabs from Unity. It can do the same thing, but they are easier to reuse and can be stacked.

**Comment by gnysek** [comment]
(╯°□°)╯︵ ┻━┻
Throw it away.
I've never used it. It's easier to make objects which change their appearance based on some variable, or event to replace tilesets, than to manage room inheritance.

**Comment by AndJamCol** [comment]
In my code library, I have a "rmEngineInit" room that creates systems objects, which is always put last in the room order every time it's re-imported into a project. To get around this, I make a child room of it, and put that child first in the room order. This is the main thing I use room inheritance for - I've tried using it for certain game rooms as well, but because the room size getting inherited creates issues it's not ideal in its current state.

**Comment by Gamer-XP** [comment]
> In my code library, I have a "rmEngineInit" room that creates systems objects, which is always put last in the room order every time it's re-imported into a project. To get around this, I make a child room of it, and put that child first in the room order. This is the main thing I use room inheritance for - I've tried using it for certain game rooms as well, but because the room size getting inherited creates issues it's not ideal in its current state.

Sounds like you're solving event execution order issue here using it. That's one way to do that instead of using persistent objects I guess.

**Comment by BackYard321** [comment]
> > In my code library, I have a "rmEngineInit" room that creates systems objects, which is always put last in the room order every time it's re-imported into a project. To get around this, I make a child room of it, and put that child first in the room order. This is the main thing I use room inheritance for - I've tried using it for certain game rooms as well, but because the room size getting inherited creates issues it's not ideal in its current state.
> 
> Sounds like you're solving event execution order issue here using it. That's one way to do that instead of using persistent objects I guess.

I'm not sure I understand what you're saying - it uses persistent objects, they just need to be created in a room first.

**Comment by Gamer-XP** [comment]
Oh, sorry. Misread that. Was thinking you was using room inheritance to store system objects as default instances, and so they are always first in the creation order.

**Comment by BackYard321** [comment]
> Oh, sorry. Misread that. Was thinking you was using room inheritance to store system objects as default instances, and so they are always first in the creation order.

Got it! Yeah I should also clarify: by "systems objects" I mean different helper objects I've set up like a cursor object, camera object, different GUI stuff, a debug controller, etc. Things that I use across a lot of different games! So having a room that creates them appearing first in the room order is fairly important for this.

**Comment by LeonardoCunhaGitHub** [comment]
Room inheritance is a must have feature.

My project relies deeply on that as it has many rooms.
It's a disaster having to update all manually every time I add or remove a layer.

It can't be that difficult for gamemaker to implement this. 

"If room is modified then update all children".

**Comment by LeonardoCunhaGitHub** [comment]
> I could nested parent rooms being useful also. For example, a level game where you might have a parent room for all the levels to set a common resolution/view. Then you could have a child room, which is parent to levels 0-9, giving them a particular theme and background, and another parent for levels 10-19 and so on. Could make development easier for these such games.

Yes that is exactly what I am doing.
I have room templates for a number of child rooms. It's very useful to have the inheritance.

Gamemaker could just fix it instead of removing it.

**Comment by muddrox** [comment]
Just lost an hour's worth of work due to a room inheritance bug.  I change my stance, just kill it, and replace it with the ability to copy and paste layers to multiple rooms at once or something because this is killing me.  I wish I never touched room inheritance.  It is the literal bane of my existence.

**Comment by LeonardoCunhaGitHub** [comment]
Agree! ... with the difference that I prefer this feature to be fixed and work properly rather then remove it because some people don't need it.

**Comment by LeonardoCunhaGitHub** [comment]
> ### Is your feature request related to a problem?
> 
> 
> 
> Room Inheritance has many fundamental issues with its functionality, complicates general development of features and the Room Editor, and there are far better ways of achieving the same results.
> 
> 
> 
> ### Describe the solution you'd like
> 
> 
> 
> Remove the functionality of room inheritance.

I cannot agree with this statement at all.
It complicates general development????
Can you please explain more in details how a room inheritance complicates your project?

An inheritance is a featured used in object orientation with the purpose of re-using the same piece of code to keep it more stable, faster and simple. It's called re-usability.

How can a game engine survive in current market without room inheritance? 
Maybe your project structure has few layers and few rooms?!!

How can a project with 10 layers and 400 rooms work without inheritance?
There is no reason to remove it...
If you don't need it just don't use it...
I think you found really useful when you found out about this feature and then after trying you noticed that is buggy but asking to remove just because is buggy seems too much...
Just don't use it.
GAMEMAKER please fix room inheritance.
It cannot be that difficult for you guys seriously.

**Comment by LeonardoCunhaGitHub** [comment]
> While I agree that room inheritance is borderline useless in its current decrepit state, I think it would be extremely useful if it were stable and complete.  It just feels unfinished in its current state.  If there are no plans to fix and improve it then sure, remove it, but I think it's a shame to lose it without anything to effectively replace it.  My vote is to fix it and improve its utility.  

Absolutely agree! 
Just fix it
Do not remove it
A game cannot go complex without that...
It's the type of feature that make people decide to go either way one engine or another...
Please fix it

**Comment by LeonardoCunhaGitHub** [comment]
> I stopped using room inheritance due to how buggy it is, and now I use a template room that I duplicate instead. However, it'd be nice to have some way to edit layers for all rooms at once, since right now if I ever want to add/remove/rename any layers, I have to go back and do it in all my existing rooms one by one. I imagine this sort of mass layer editing could be added in a way that's not inheritance, so I'd like to see that. Something like how you can select multiple assets and edit properties for all of them at once in the inspector, but for rooms

Just select the layer or room and turn off room inheritance.

**Comment by LeonardoCunhaGitHub** [comment]
> As others are saying, the current implementation of room inheritance is unreliable.
> 
> That being said, I believe the concept is close to gold. There are many occasions where I'd like have a common view size across all my rooms, perhaps a background layer or in the future, a shared GUI layer. This means that any change to the parent room, will be inherited to all the child rooms, allowing for resolution and other changes to be done quickly. This is what I'd like to see from room inheritance. A simple way to change parameters across multiple rooms at once.
> 
> I could nested parent rooms being useful also. For example, a level game where you might have a parent room for all the levels to set a common resolution/view. Then you could have a child room, which is parent to levels 0-9, giving them a particular theme and background, and another parent for levels 10-19 and so on. Could make development easier for these such games.

Agree. This is what I am telling everybody... they want to remove it because their project or implementation strategy does not require inheritance but I keep saying... if you don't want it then don't use it.
I cannot imagine a developer working with rooms without inheritance.

**Comment by LeonardoCunhaGitHub** [comment]
> Current thoughts are too leave it (Room Inheritance) in the current room editor but in the New Room Editor (which we are designing / thinking about now) we are thinking about removing it completely.
> 
> 
> 
> You would be able to use the current room editor for your current project.
> 
> 
... and what happens to the developers who started their project years ago using room inheritance?
They will have to use the older GM version?
Does that sound fair to you?
I want to use the newest version and be assured that the features won't start disappearing from the engine.

I started my project 5 years ago and I use room inheritance heavily on 400 rooms.

So removing it will affect me a lot.
I am pretty sure many other also that are not even reading this thread.

**Comment by LeonardoCunhaGitHub** [comment]
> ### Is your feature request related to a problem?
> 
> 
> 
> Room Inheritance has many fundamental issues with its functionality, complicates general development of features and the Room Editor, and there are far better ways of achieving the same results.
> 
> 
> 
> ### Describe the solution you'd like
> 
> 
> 
> Remove the functionality of room inheritance.

**Describe the solution you'd like**

Fix room inheritance properly.

**Comment by paxer** [comment]
I would prefer to have room inheritance to be fixed and work properly rather than removed. It is a useful feature for many cases.

**Comment by gnysek** [comment]
Since `2023.11` introduces `room_get_info`, "parent" room instances can be easily copied from ANY room, including Variable Definitions override and Creation code. Same for tilesets and any other elements.

I suspect that before release of `2023.11` we gonna have community library/scripts which does that.

**Comment by LeonardoCunhaGitHub** [comment]
Sorry not sure I understand your point of view.
I believe commands work during execution time and the room inheritance needs to be manageable not only by code but mostly importantly on room editor level.
Right?

**Comment by AtlaStar** [comment]
> Sorry not sure I understand your point of view. I believe commands work during execution time and the room inheritance needs to be manageable not only by code but mostly importantly on room editor level. Right?

You can run code in script assets, which run before room creation code...dunno if the room info exists at that point yet or not though

**Comment by yoshikins** [comment]
Never once used room inheritence because it caused more problems than it solved. I always make a template room in my projects and duplicate that room.

**Comment by BackYard321** [comment]
> In my code library, I have a "rmEngineInit" room that creates systems objects, which is always put last in the room order every time it's re-imported into a project. To get around this, I make a child room of it, and put that child first in the room order. This is the main thing I use room inheritance for - I've tried using it for certain game rooms as well, but because the room size getting inherited creates issues it's not ideal in its current state.

Somewhat related to what @AtlaStar mentioned - I realized I can create a script asset with `room_goto(rmEngineInit)`, which eliminates the need for a child room in this case!

**Comment by BackYard321** [comment]
> > In my code library, I have a "rmEngineInit" room that creates systems objects, which is always put last in the room order every time it's re-imported into a project. To get around this, I make a child room of it, and put that child first in the room order. This is the main thing I use room inheritance for - I've tried using it for certain game rooms as well, but because the room size getting inherited creates issues it's not ideal in its current state.
> 
> Somewhat related to what @AtlaStar mentioned - I realized I can create a script asset with `room_goto(rmEngineInit)`, which eliminates the need for a child room in this case!

Nevermind; this won't work when you call `room_restart()`. So yeah, if there's something that would respect room order when packages are re-imported, that would be great!

**Comment by AvioxArcade** [comment]
I used room inheritance in a project that required certain Layers (with specific names) to exist in each room so they could be processed properly when loading a room. 

An example of how this played out -- I had an autotile layer that I used to draw a brush/bushlike boundaries in a top-down game. 
Using an autotile layer made iterating (and initial creation) of these rooms much faster, but I needed them to both have collision and depth if standing above/behind them.  So I would run code at a `room_start` event that would check for a layer with a specific name and convert relevant tiles into objects with depth or objects with collision. 

I did similar things for other types of environment, e.g. water or cliffs.

I realize there are other ways to accomplish this, but that was my use-case.

Using inheritance often broke and messed up depths, etc, when making changes, so it definitely was a pain to use.
